scalanet/src/io/iohk/scalanet/peergroup/kademlia/KRouter.scala

```diff
@@ -354,46 +362,50 @@ class KRouter[A](
       }
     }
 
-    closestKnownNodesTask.flatMap { closestKnownNodes =>
-      if (closestKnownNodes.isEmpty) {
-        Task(logger.debug("Lookup finished without any nodes, as bootstrap nodes ")).as(Seq.empty)
-      } else {
-        val initalRequestState = closestKnownNodes.foldLeft(Map.empty[BitVector, RequestResult]) { (map, node) =>
-          map + (node.id -> RequestScheduled)
-        }
-        // All initial nodes are scheduled to request
-        val lookUpTask: Task[Seq[NodeRecord[A]]] = for {
+    closestKnownNodesTask
+      .map {
+        case h :: t => NonEmptySet.of(h, t: _*).some
+        case Nil => none
+      }
+      .flatMap {
+        case None =>
+          Task(logger.debug("Lookup finished without any nodes, as bootstrap nodes ")).as(SortedSet.empty)
+
+        case Some(closestKnownNodes) =>
+          val initalRequestState: Map[BitVector, RequestResult] =
+            closestKnownNodes.toList.map(_.id -> RequestScheduled).toMap
           // All initial nodes are scheduled to request
-          state <- Ref.of[Task, Map[BitVector, RequestResult]](initalRequestState)
-          // closestKnownNodes are constrained by alpha, it means there will be at most alpha independent recursive tasks
-          results <- Task.parTraverse(closestKnownNodes) { knownNode =>
-            recLookUp(List(knownNode), NonEmptyList.fromListUnsafe(closestKnownNodes.toList), state)
-          }
-          records = results.flatMap(_.toList)
-        } yield records
+          val lookUpTask: Task[SortedSet[NodeRecord[A]]] = for {
+            // All initial nodes are scheduled to request
+            state <- Ref.of[Task, Map[BitVector, RequestResult]](initalRequestState)
+            // closestKnownNodes are constrained by alpha, it means there will be at most alpha independent recursive tasks
+            results <- Task.parTraverse(closestKnownNodes.toList) { knownNode =>
```

Comment:
thic could probably be: `parTraverseUnordered`, also (https://github.com/input-output-hk/scalanet/pull/85/files#diff-3efccc289265a0426b8121d2294a1b1bR342)